mage/src/mage/modules.clj

```diff
@@ -0,0 +1,163 @@
+
+;;; TODO (Cam 2025-11-07) changes to test files should only cause us to run tests for that module as well, not
+;;; everything that depends on that module directly or indirectly in `src`
+(defn- file->module [filename]
+  (or
+   (when-let [[_match module] (re-matches #"^(?:(?:src)|(?:test))/metabase/([^/]+)/.*$" filename)]
+     (symbol (str/replace module #"_" "-")))
+   (when-let [[_match module] (re-matches #"^enterprise/backend/(?:(?:src)|(?:test))/metabase_enterprise/([^/]+)/.*$" filename)]
+     (symbol "enterprise" (str/replace module #"_" "-")))))
+
+(defn- updated-files->updated-modules [updated-files]
+  (into (sorted-set)
+        (keep file->module)
+        updated-files))
+
+(defn- updated-modules [git-ref]
+  (let [git-ref       (or git-ref "master")
+        updated-files (u/updated-files git-ref)]
+    (updated-files->updated-modules updated-files)))
+
+(defn- module->test-directory
+  [module]
+  (case (namespace module)
+    "enterprise" (str "enterprise/backend/test/metabase_enterprise/" (str/replace (name module) #"-" "_"))
+    nil          (str "test/metabase/" (str/replace (name module) #"-" "_"))))
+
+(defn- dependencies
+  "Read out the Kondo config for the modules linter; return a map of module => set of modules it directly depends on."
+  []
+  (let [config (-> (with-open [r (java.io.PushbackReader. (java.io.FileReader. ".clj-kondo/config/modules/config.edn"))]
+                     (edn/read r))
+                   :metabase/modules
+                   ;; ignore the config for [[metabase.connection-pool]] which comes from one of our libraries.
+                   (dissoc 'connection-pool))]
+    (into (sorted-map)
+          (map (fn [[k config]]
+                 [k (:uses config)]))
+          config)))
+
+(defn- direct-dependents
+  "Set of modules that directly depend on `module`."
+  [deps module]
+  (into (sorted-set)
+        (keep (fn [[a-module module-deps]]
+                (when (or (= module-deps :any)
+                          (contains? module-deps module))
+                  a-module)))
+        deps))
+
+(comment
+  (direct-dependents (dependencies) 'driver))
+
+(defn- indirect-dependents
+  "Set of modules that either directly or indirectly depend on `module`."
+  ([deps module]
+   (indirect-dependents deps module (sorted-set)))
+  ([deps module acc]
+   (let [module-deps (direct-dependents deps module)
+         new-deps    (set/difference module-deps acc)
+         acc         (into acc new-deps)]
+     (reduce
+      (fn [acc new-dep]
+        (indirect-dependents deps new-dep acc))
+      acc
+      new-deps))))
+
+(defn- affected-modules
+  "Set of modules that are direct or indirect dependents of `modules`, and thus are affected by changes to them."
+  [deps modules]
+  (into (sorted-set)
+        (mapcat (partial indirect-dependents deps))
+        modules))
+
+(defn- unaffected-modules
+  "Return the set of modules that are unaffected "
+  [deps modules]
+  (set/difference
```

Comment:
we probably want to put a `(into (sorted-set) ` around the outer `set/difference` here so that when we print the unaffected modules they are sorted. sorting inside of the `set/difference` seems to not be helpful.